pkg/agent/openflow/pipeline.go

```diff
@@ -2545,70 +2593,91 @@ func (c *client) genPacketInMeter(meterID binding.MeterIDType, rate uint32) bind
-	c.createOFTable(L2ForwardingCalcTable, IngressMetricTable.GetID(), binding.TableMissActionNext)
-	c.createOFTable(IngressRuleTable, IngressDefaultTable.GetID(), binding.TableMissActionNext)
-	c.createOFTable(IngressDefaultTable, IngressMetricTable.GetID(), binding.TableMissActionNext)
-	c.createOFTable(IngressMetricTable, ConntrackCommitTable.GetID(), binding.TableMissActionNext)
-	c.createOFTable(ConntrackCommitTable, L2ForwardingOutTable.GetID(), binding.TableMissActionNext)
-	c.createOFTable(L2ForwardingOutTable, binding.LastTableID, binding.TableMissActionDrop)
-	if c.enableProxy {
-		SpoofGuardTable.SetNext(ServiceHairpinTable.GetID())
-		IPv6Table.SetNext(ServiceHairpinTable.GetID())
-		ConntrackStateTable.SetNext(EndpointDNATTable.GetID())
-		if c.proxyAll {
-			ServiceHairpinTable = c.createOFTable(ServiceHairpinTable, ServiceConntrackTable.GetID(), binding.TableMissActionNext)
-			ServiceConntrackTable = c.createOFTable(ServiceConntrackTable, ConntrackTable.GetID(), binding.TableMissActionNext)
-			ServiceClassifierTable = c.createOFTable(ServiceClassifierTable, binding.LastTableID, binding.TableMissActionNone)
-			ServiceConntrackCommitTable = c.createOFTable(ServiceConntrackCommitTable, HairpinSNATTable.GetID(), binding.TableMissActionNext)
-		} else {
-			ServiceHairpinTable = c.createOFTable(ServiceHairpinTable, ConntrackTable.GetID(), binding.TableMissActionNext)
-		}
-		SessionAffinityTable = c.createOFTable(SessionAffinityTable, binding.LastTableID, binding.TableMissActionNone)
-		ServiceLBTable = c.createOFTable(ServiceLBTable, EndpointDNATTable.GetID(), binding.TableMissActionNext)
-		EndpointDNATTable = c.createOFTable(EndpointDNATTable, c.egressEntryTable, binding.TableMissActionNext)
-		HairpinSNATTable = c.createOFTable(HairpinSNATTable, L2ForwardingOutTable.GetID(), binding.TableMissActionNext)
-		ConntrackCommitTable.SetNext(HairpinSNATTable.GetID())
-	} else {
-		c.createOFTable(DNATTable, c.egressEntryTable, binding.TableMissActionNext)
-	}
-	// The default SNAT is implemented with OVS on Windows.
-	if c.enableEgress || runtime.IsWindowsPlatform() {
-		c.createOFTable(SNATTable, L2ForwardingCalcTable.GetID(), binding.TableMissActionNext)
-	}
-	if runtime.IsWindowsPlatform() || c.connectUplinkToBridge {
-		c.createOFTable(UplinkTable, SpoofGuardTable.GetID(), binding.TableMissActionNone)
+func generatePipeline(templates []*featureTemplate) binding.Pipeline {
+	pipelineID := binding.NewPipelineID()
+	sortedOFTableMap := make(map[binding.StageID][]binding.Table)
+
+	// PipelineClassifierTable ID is always 0, and it is the first table for all pipelines. Create PipelineClassifierTable
+	// on the bridge when building a pipeline.
+	if PipelineClassifierTable.ofTable == nil {
+		PipelineClassifierTable.ofTable = binding.NewOFTable(binding.NextTableID(), PipelineClassifierTable.name, 0, binding.AllPipelines)
 	}
```

Comment:
do you think it would make sense to have a method on the `featureTable` type to instantiate the `ofTable` member?

Reply:
I think it's ok. IMO, it makes no difference to add such method.